cmd/scollector/collectors/haproxy_unix.go

```diff
@@ -0,0 +1,528 @@
+func AddHAProxyInstance(conf string) error {
+	sp := strings.SplitN(conf, ":", 2)
+	if len(sp) < 2 {
+		return fmt.Errorf("haproxy_instance requires two fields")
+	}
+	var instance trackedInstance
+	instance.Tier = sp[0]
+	instance.URL = sp[1]
+	trackedInstances = append(trackedInstances, &instance)
+
+	return nil
+}
+
+func cHAProxyCSV(user, pwd string) (opentsdb.MultiDataPoint, error) {
+	var md opentsdb.MultiDataPoint
+	var err error
+
+	for _, instance := range trackedInstances {
+		if e := instance.fetchHAProxyData(&md, user, pwd); e != nil {
+			err = e
+		}
+	}
+	return md, err
+}
+
+func (instance *trackedInstance) fetchHAProxyData(md *opentsdb.MultiDataPoint, user string, pwd string) error {
+	var err error
+	const metric = "haproxy"
+	parse := func(v string) (int64, error) {
+		var i int64
+		if v != "" {
+			i, err = strconv.ParseInt(v, 10, 64)
+			if err != nil {
+				return 0, err
+			}
+			return i, nil
+		}
+		return i, nil
+	}
+
+	req, err := http.NewRequest("GET", instance.URL, nil)
+	if err != nil {
+		return err
+	}
+	req.SetBasicAuth(user, pwd)
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	reader := csv.NewReader(resp.Body)
+	records, err := reader.ReadAll()
+	if err != nil {
+		return err
+	}
+	if len(records) < 2 {
+		return nil
+	}
+	for _, rec := range records[1:] {
+		// There is a trailing comma in haproxy's csv
+		if len(rec) != len(haproxyCSVMeta)+1 {
+			return fmt.Errorf("expected %v lines. got: %v",
+				len(haproxyCSVMeta)+1, len(rec))
+		}
+		hType := haproxyType[rec[32]]
+		pxname := rec[0]
+		svname := rec[1]
+		ts := opentsdb.TagSet{"pxname": pxname, "svname": svname, "tier": instance.Tier}
+		for i, field := range haproxyCSVMeta {
+			m := strings.Join([]string{metric, hType, field.Name}, ".")
+			value := rec[i]
+			if field.Ignore == true {
+				continue
+			} else if strings.HasPrefix(field.Name, "hrsp") {
+				sp := strings.Split(field.Name, "_")
+				if len(sp) != 2 {
+					return fmt.Errorf("unexpected field name %v in hrsp", field.Name)
+				}
+				ts := ts.Copy().Merge(opentsdb.TagSet{"status_code": sp[1]})
+				m = strings.Join([]string{metric, hType, sp[0]}, ".")
+				v, err := parse(value)
+				if err != nil {
+					return err
+				}
+				Add(md, m, v, ts, metadata.Counter, metadata.Response,
+					fmt.Sprintf("The number of http responses with a %v status code.", sp[1]))
+			} else if field.Name == "status" {
+				v, ok := haproxyStatus[value]
+				// Not distinging between MAINT and MAINT via...
+				if !ok {
+					v = 3
+				}
+				Add(md, m, v, ts, field.RateType, field.Unit, field.Desc)
+			} else if field.Name == "check_status" {
+				if value == "" {
+					continue
+				}
+				v, ok := haproxyCheckStatus[value]
+				if !ok {
+					return fmt.Errorf("unknown check status %v", value)
+				}
+				Add(md, m, v, ts, field.RateType, field.Unit, field.Desc)
+			} else {
+				v, err := parse(value)
+				if err != nil {
+					return err
```

Comment:
I believe the intent there is to throw an error and stop processing. If any field fails to parse, then something odd is happening somewhere.